drivers/ethernet/oa_tc6.c

```diff
@@ -306,202 +690,201 @@ int oa_tc6_check_status(struct oa_tc6 *tc6)
+				break;
+			}
+		}
+	}
 }
 
-int oa_tc6_read_status(struct oa_tc6 *tc6, uint32_t *ftr)
+static void oa_tc6_int_callback(const struct device *dev, struct gpio_callback *cb, uint32_t pins)
 {
-	uint32_t hdr;
+	ARG_UNUSED(dev);
+	ARG_UNUSED(pins);
 
-	hdr = FIELD_PREP(OA_DATA_HDR_DNC, 1) | FIELD_PREP(OA_DATA_HDR_DV, 0) |
-	      FIELD_PREP(OA_DATA_HDR_NORX, 1);
-	hdr |= FIELD_PREP(OA_DATA_HDR_P, oa_tc6_get_parity(hdr));
+	struct oa_tc6 *tc6 = CONTAINER_OF(cb, struct oa_tc6, gpio_int_callback);
 
-	return oa_tc6_chunk_spi_transfer(tc6, NULL, NULL, hdr, ftr);
+	if (!tc6->rst_flag) {
+		k_sem_give(&tc6->int_sem);
+	} else {
+		tc6->int_flag = true;
+		k_sem_give(&tc6->spi_sem);
+	}
 }
 
-int oa_tc6_read_chunks(struct oa_tc6 *tc6, struct net_pkt *pkt)
+static void oa_tc6_calculate_spi_buffer_size(struct oa_tc6 *tc6)
 {
-	const uint16_t buf_rx_size = CONFIG_NET_BUF_DATA_SIZE;
-	struct net_buf *buf_rx = NULL;
-	uint32_t buf_rx_used = 0;
-	uint32_t hdr, ftr;
-	uint8_t sbo, ebo;
-	int ret;
+	tc6->oa_spi_tx_rx_buffer_size = CONFIG_OA_TC6_TX_RX_BUFFER_SIZE;
 
 	/*
-	 * Special case - append already received data (extracted from previous
-	 * chunk) to new packet.
+	 * OA SPI TX RX buffer size value should be multiples of 68 (size of the single OA SPI
+	 * chunk is 68 bytes)
 	 *
-	 * This code is NOT used when OA_CONFIG0 RFA [13:12] is set to 01
-	 * (ZAREFE) - so received ethernet frames will always start on the
-	 * beginning of new chunks.
+	 * If not, calculate the nearest mimimum multiple value
 	 */
-	if (tc6->concat_buf) {
-		net_pkt_append_buffer(pkt, tc6->concat_buf);
-		tc6->concat_buf = NULL;
+	if (CONFIG_OA_TC6_TX_RX_BUFFER_SIZE % tc6->chunk_size != 0) {
+		tc6->oa_spi_tx_rx_buffer_size =
+			(CONFIG_OA_TC6_TX_RX_BUFFER_SIZE / tc6->chunk_size) * tc6->chunk_size;
 	}
+}
 
-	do {
-		if (!buf_rx) {
-			buf_rx = net_pkt_get_frag(pkt, buf_rx_size, OA_TC6_BUF_ALLOC_TIMEOUT);
-			if (!buf_rx) {
-				LOG_ERR("OA RX: Can't allocate RX buffer fordata!");
-				return -ENOMEM;
-			}
-		}
+int oa_tc6_init(struct oa_tc6 *tc6)
+{
+	int ret;
 
-		hdr = FIELD_PREP(OA_DATA_HDR_DNC, 1);
-		hdr |= FIELD_PREP(OA_DATA_HDR_P, oa_tc6_get_parity(hdr));
+	tc6->chunk_size = tc6->cps + OA_TC6_HDR_SIZE;
 
-		ret = oa_tc6_chunk_spi_transfer(tc6, buf_rx->data + buf_rx_used, NULL, hdr, &ftr);
-		if (ret < 0) {
-			LOG_ERR("OA RX: transmission error: %d!", ret);
-			goto unref_buf;
-		}
+	k_sem_init(&tc6->spi_sem, 0, 1);
+	k_sem_init(&tc6->tx_enq_sem, 0, 1);
 
-		ret = -EIO;
-		if (oa_tc6_get_parity(ftr)) {
-			LOG_ERR("OA RX: Footer parity error!");
-			goto unref_buf;
-		}
+	tc6->waiting_net_pkt = NULL;
+	tc6->ongoing_net_pkt = NULL;
 
-		if (!FIELD_GET(OA_DATA_FTR_SYNC, ftr)) {
-			LOG_ERR("OA RX: Configuration not SYNC'ed!");
-			goto unref_buf;
-		}
+	oa_tc6_calculate_spi_buffer_size(tc6);
 
-		if (!FIELD_GET(OA_DATA_FTR_DV, ftr)) {
-			LOG_DBG("OA RX: Data chunk not valid, skip!");
-			goto unref_buf;
-		}
+#ifdef CONFIG_SPI_ASYNC
```

Comment:
I'm wondering if this code (i.e. without semaphores enabled) can work reliably?
Apparently for my setup (STM32F + LAN865x) it has been implicitly assumed that `CONFIG_SPI_ASYNC` is defined.

Reply:
@lmajewski  SAM0 devices like SAM E54 XPro spi driver (spi_sam0.c) implemented DMA support in the Asynchronous call support. So it is necessary to enable Asynchronous call support to achieve the maximum performance with DMA support. Whereas STM32 spi driver (spi_ll_stm32.c) implemented DMA support in the synchronous call. So enabling Asynchronous call support in the stm32 platform will limit the performance.

To ensure that CONFIG_SPI_ASYNC code block is executed only for SAM0 devices, I can change the code like below:
```suggestion
#if defined(CONFIG_SPI_ASYNC) && (defined(CONFIG_SOC_SERIES_SAMC21) || 
		                  defined(CONFIG_SOC_SERIES_SAMD20) || 
				  defined(CONFIG_SOC_SERIES_SAMD21) ||
				  defined(CONFIG_SOC_SERIES_SAME54) ||
				  defined(CONFIG_SOC_SERIES_SAML21) ||
				  defined(CONFIG_SOC_SERIES_SAMR21) ||
				  defined(CONFIG_SOC_SERIES_SAMR34))
```

Reply:
At least for mine `stm32g4xx` I do have:
`#CONFIG_SPI_STM32_DMA`
`#CONFIG_SPI_ASYNC`
`CONFIG_SPI_STM32=y`
`CONFIG_SPI_STM32_USE_HW_SS=y`

so no DMA and no asynchronous operation is supported.
@ABOSTM @erwango - I'm wondering if such support is planned in the stm's HAL? Or do I miss something?

This is anyway a side discussion as the driver shall work reliably with any SPI API supported.

Reply:
if you want to optimize, that should be a seperate Kconfig, that enables spi async just for this driver, with a dependency on CONFIG_SPI_ASYNC and a  default y if for the fast soc with async dma